README.md
Cheerp: A C++ compiler for the Web
==================================

Please report bugs on launchpad:
https://bugs.launchpad.net/cheerp

Cheerp GLES implementation installation
---------------------------------------

```
make -C webgles install INSTALL_PREFIX=/opt/cheerp CHEERP_PREFIX=/opt/cheerp
```

Cheerp helper library for Wasm
---------------------------------------

```
make -C wasm install INSTALL_PREFIX=/opt/cheerp CHEERP_PREFIX=/opt/cheerp
```

Cheerp PreExecuted standard C++ library installation (both generic and asm.js version)
----------------------------------------------------

```
make -C stdlibs install INSTALL_PREFIX=/opt/cheerp CHEERP_PREFIX=/opt/cheerp
```

Cheerp PreExecuted standard C++ library installation (only generic js version)
----------------------------------------------------

```
make -C stdlibs install_genericjs INSTALL_PREFIX=/opt/cheerp CHEERP_PREFIX=/opt/cheerp
```

Cheerp PreExecuted standard C++ library installation (only asm.js version)
----------------------------------------------------

```
make -C stdlibs install_asmjs INSTALL_PREFIX=/opt/cheerp CHEERP_PREFIX=/opt/cheerp
```



debian/changelog
cheerp-libs (2.6-1ubuntu1) focal; urgency=medium

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Mon, 30 Nov 2020 10:57:13 +0100

cheerp-libs (2.5-1ubuntu1) bionic; urgency=medium

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Sat, 04 Apr 2020 18:35:26 +0200

cheerp-libs (2.0-1ubuntu1) bionic; urgency=medium

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Wed, 13 Feb 2019 14:55:00 +0100

cheerp-libs (2.0-rc2-1) artful; urgency=medium

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Mon, 23 Apr 2018 15:01:54 +0200

cheerp-libs (2.0-rc1-1) xenial; urgency=medium

  * New release

 -- Sander Mathijs van Veen <sander@leaningtech.com>  Tue, 24 Oct 2017 11:30:34 +0200

cheerp-libs (1.3-1) xenial; urgency=medium

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Tue, 27 Sep 2016 09:40:35 +0200

cheerp-libs (1.2-1) wily; urgency=medium

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Thu, 21 Jan 2016 11:16:15 +0100

cheerp-libs (1.1-1) utopic; urgency=medium

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Wed, 03 Jun 2015 10:44:48 +0200

cheerp-libs (1.0-1) trusty; urgency=medium

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Wed, 21 May 2014 16:33:37 +0200

cheerp-libs (0.9.6-1) saucy; urgency=low

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Thu, 17 Apr 2014 17:27:01 +0200

cheerp-libs (0.9.5-1) saucy; urgency=low

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Thu, 03 Apr 2014 16:21:47 +0200

cheerp-libs (0.9.4-1) saucy; urgency=low

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Mon, 17 Mar 2014 02:07:56 +0100

cheerp-libs (0.9.3-1) saucy; urgency=low

  * New release

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Thu, 02 Jan 2014 18:32:15 +0100

cheerp-libs (0.9.2-1) saucy; urgency=low

  * Debian fixes

 -- Alessandro Pignotti <alessandro@leaningtech.com>  Sat, 30 Nov 2013 20:29:55 +0100

cheerp-libs (0.9.2) raring; urgency=low

  * Initial release.

 -- Alessandro Pignotti <a.pignotti@sssup.it>  Thu, 26 Nov 2013 22:56:40 +0100


wasm/memory.cpp
/****************************************************************
 *
 * Copyright (C) 2018 Alessandro Pignotti <alessandro@leaningtech.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 ***************************************************************/

#include <stddef.h>

extern "C"
{
	/** Implement a faster variant of memcpy, specialized for wasm
	 *  At link time this will replace the 'weak' symbol for memcpy
	 *  provided in the lib C
	 */
	void* memcpy(void* const dst, void* const src, size_t len)
	{
		// Wasm supports unaligned load/stores, take advantage of this for a faster memcpy
		unsigned char* src8 = (unsigned char*)src;
		unsigned char* dst8 = (unsigned char*)dst;

		unsigned char* srcEnd = src8+len;

		while(((unsigned long)dst8)&3)
		{
			if(src8 == srcEnd)
				return dst;
			*dst8 = *src8;
			dst8++;
			src8++;
		}
		// Now the dest pointer is aligned
		// Unroll for 64 bytes at a time
		while(int(src8) <= int(srcEnd - 64))
		{
			*((unsigned long long*)dst8) = *((unsigned long long*)src8);
			*((unsigned long long*)(dst8+8)) = *((unsigned long long*)(src8+8));
			*((unsigned long long*)(dst8+16)) = *((unsigned long long*)(src8+16));
			*((unsigned long long*)(dst8+24)) = *((unsigned long long*)(src8+24));
			*((unsigned long long*)(dst8+32)) = *((unsigned long long*)(src8+32));
			*((unsigned long long*)(dst8+40)) = *((unsigned long long*)(src8+40));
			*((unsigned long long*)(dst8+48)) = *((unsigned long long*)(src8+48));
			*((unsigned long long*)(dst8+56)) = *((unsigned long long*)(src8+56));
			dst8+=64;
			src8+=64;
		}
		// Loop 8 bytes at a time
		while(int(src8) <= int(srcEnd - 8))
		{
			*((unsigned long long*)dst8) = *((unsigned long long*)src8);
			dst8+=8;
			src8+=8;
		}
		// Byte loop to finish the copy
		while(src8 != srcEnd)
		{
			*dst8 = *src8;
			dst8++;
			src8++;
		}
		return dst;
	}

	void *memset(void *dst, int c, size_t len)
	{
		unsigned char cu = c;
		unsigned int c32 = (cu<<0) | (cu<<8) | (cu<<16) | (cu<<24);
		unsigned long long c64 = ((unsigned long long)(c32)<<0) | ((unsigned long long)(c32)<<32);
		unsigned char* dst8 = (unsigned char*)dst;
		unsigned char* dstEnd = dst8 + len;
		while(int(dst8) < int(dstEnd - 64))
		{
			*((unsigned long long*)dst8) = c64;
			*((unsigned long long*)(dst8+8)) = c64;
			*((unsigned long long*)(dst8+16)) = c64;
			*((unsigned long long*)(dst8+24)) = c64;
			*((unsigned long long*)(dst8+32)) = c64;
			*((unsigned long long*)(dst8+40)) = c64;
			*((unsigned long long*)(dst8+48)) = c64;
			*((unsigned long long*)(dst8+56)) = c64;
			dst8+=64;
		}
		// Loop 8 bytes at a time
		while(int(dst8) <= int(dstEnd - 8))
		{
			*((unsigned long long*)dst8) = c64;
			dst8+=8;
		}
		// Byte loop to finish the copy
		while(dst8 != dstEnd)
		{
			*dst8 = cu;
			dst8++;
		}
		return dst;
	}
}
